### PR TITLE
Disallow `TypeAlias` type

### PR DESCRIPTION
**Pull request by sobolevn**
I've created two functions `is_forbidden_to_register` and `is_forbidden_to_dispatch`.
I am planning to refactor them to match our rules. For example, we can dispatch `Annotated` type, but it does not make sense with `.register_type_strategy`.

Refs https://github.com/HypothesisWorks/hypothesis/issues/2978

### CODE REVIEW

**Comment by sobolevn** [comment]
Btw, I am pretty sure that Catholic Christmas is celebrated in Australia, so merry cristmas to @Zac-HD 🌲

**Comment by Zac-HD** [comment]
> Btw, I am pretty sure that Catholic Christmas is celebrated in Australia, so merry cristmas to @Zac-HD 🌲

Thanks 💖  Yeah, we celebrate on (or around 😅) Dec. 25th, so I've just had two huge lunches with extended family.  The entire country basically shuts down for a holiday from ~20 Dec to ~5 Jan (or ~28 Jan, after Australia Day, for school holidays!) -- it's pretty similar to July in Europe or Canada.  [*White Wine in the Sun*](https://www.youtube.com/watch?v=fCNvZqpa-7Q) captures the mood pretty well, especially since I'm moving in a few weeks...

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py [hunk 1]
```diff
@@ -57,6 +57,14 @@
     except ImportError:
         _AnnotatedAlias = ()
 
+try:
+    from typing import TypeAlias  # type: ignore
+except ImportError:
+    try:
+        from typing_extensions import TypeAlias
+    except ImportError:
+        TypeAlias = object()  # It must be a singleton
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/types.py:66]: Tricky edge case here: we want to handle `typing_extensions.TypeAlias`, even when `typing.TypeAlias` exists, so I don't think we need this import logic.

(Since the tests passs it looks like this is just ~~an alias~~ re-exported from `typing` if available, but I'd rather not assume that in our implementation)

File: hypothesis-python/RELEASE.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+RELEASE_TYPE: minor
+
+This release disallows using :obj:`python:typing.TypeAlias`
+with :func:`~hypothesis.strategies.from_type`
+and :func:`~hypothesis.strategies.register_type_strategy`.
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst:5]: Let's also say *why* here, and reference `:issue:2978` for details so that users can chime in if it affects them.

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py [hunk 3]
```diff
@@ -1023,6 +1023,9 @@ def as_strategy(strat_or_callable, thing, final=True):
                 "strings."
             )
         raise InvalidArgument(f"thing={thing!r} must be a type")  # pragma: no cover
+    if types.is_forbidden_to_dispatch(thing):  # pragma: no cover
+        # Code like `st.from_type(TypeAlias)` does not make sense.
+        raise InvalidArgument(f"thing={thing!r} does not make sense as a strategy")
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:1028]: This is going to be a bit awkward, but I think I'd prefer to write a separate clause for each inline, so that we can have a detailed and specific error message.

I can imagine wanting abstraction or helper functions when we're handling 4+ types like this, but it feels premature now.

**Review by Zac-HD — COMMENTED** [review]
Basically looks good to me 😁 

...fiddly comments below though 😅

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py (hunk 1 above)
Comment (sobolevn) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py:66]: Good catch 👍

**Comment by sobolevn** [comment]
> Yeah, we celebrate on (or around 😅) Dec. 25th, so I've just had two huge lunches with extended family. The entire country basically shuts down for a holiday from ~20 Dec to ~5 Jan

We are going to start celebrating New Year and then go for two full weeks of holidays, including [Ortodox Cristmas on 7th of Jan](https://en.wikipedia.org/wiki/Christmas_in_Russia), we stop thinking "it is still holidays" on ["Old New Year" day](https://en.wikipedia.org/wiki/Old_New_Year). One of the most favourite times of the year! ❄️ 

I will fix all comments later today, thanks a lot for the review! 👍

**Comment by sobolevn** [comment]
@Zac-HD can you please take a look? https://github.com/HypothesisWorks/hypothesis/runs/4657300352?check_suite_focus=true This does not seem related. And `hypothesis.extra.redis.RedisExampleDatabase` seems to still be there.

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py [hunk 4]
```diff
@@ -57,6 +57,20 @@
...
+except ImportError:
+    pass  # Is missing for `python<3.10`
+else:  # pragma: no cover
+    TypeAliasTypes += (TypingTypeAlias,)
+try:
+    from typing_extensions import TypeAlias as ExtensionsTypeAlias
+except ImportError:
+    pass  # Is missing for `typing_extensions<3.10`
+else:  # pragma: no cover
+    TypeAliasTypes += (ExtensionsTypeAlias,)
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/types.py:72]:
```suggestion
try:
    TypeAliasTypes += (typing_extensions.TypeAlias,)
except AttributeError:
    pass  # typing_extensions package not available, or older than v3.10
```

And similar for `typing`, which we always import (but might raise `AttributeError`)

File: hypothesis-python/docs/changes.rst [hunk 5]
```diff
@@ -18,6 +18,23 @@ Hypothesis 6.x
 
     .. include:: ../RELEASE.rst
 
+.. _v6.34.0:
+
+-------------------
+6.34.0 - 2021-12-29
+-------------------
```
Comment (Zac-HD) [inline hypothesis-python/docs/changes.rst:25]: Looks like you accidentally committed the new changelog!  Best to revert this (and the version number bump); it should be handled by our CI infra.

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py [hunk 6]
```diff
@@ -1023,6 +1023,9 @@ def as_strategy(strat_or_callable, thing, final=True):
                 "strings."
             )
         raise InvalidArgument(f"thing={thing!r} must be a type")  # pragma: no cover
+    if thing in types.TypeAliasTypes:  # pragma: no cover
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:1026]: Let's add the `typing_extensions` tests to our coverage runs.  It's a small perf hit, but I'll be happier when I can see that these branches are definitely covered 😁

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py [hunk 7]
```diff
@@ -1023,6 +1023,9 @@ def as_strategy(strat_or_callable, thing, final=True):
...
+        # Code like `st.from_type(TypeAlias)` does not make sense.
+        raise InvalidArgument(f"thing={thing!r} does not make sense as a strategy")
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:1028]: Maybe a message like "Cannot resolve TypeAlias to a strategy, because there are no instances of it at runtime"?

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py [hunk 8]
```diff
@@ -1751,6 +1754,10 @@ def register_type_strategy(
 
     if not types.is_a_type(custom_type):
         raise InvalidArgument(f"custom_type={custom_type!r} must be a type")
+    elif custom_type in types.TypeAliasTypes:  # pragma: no cover
+        raise InvalidArgument(
+            f"custom_type={custom_type!r} is not allowed to be registered"
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:1759]: "... because there is no such thing as a runtime instance of TypeAlias"

File: hypothesis-python/tests/typing_extensions/test_backported_types.py [hunk 9]
```diff
@@ -126,3 +137,24 @@ def test_annotated_with_two_strategies(data):
 @given(st.data())
 def test_annotated_extra_metadata(data):
     assert data.draw(st.from_type(ExtraAnnotationNoStrategy)) > 0
+
+
+@pytest.mark.parametrize(
+    "type_alias_type",
+    [
+        TypeAlias,  # It is always available from `typing_extensions`
```
Comment (Zac-HD) [inline hypothesis-python/tests/typing_extensions/test_backported_types.py:145]: "from recent versions of typing_extensions"

**Review by Zac-HD — COMMENTED** [review]
It looks like the `redis` warning in our docs build is because they moved the docs upstream!  See https://github.com/redis/redis-py/issues/1839; if this doesn't get fixed upstream we'll need to add a `.client` into the relevant docstring in `hypothesis.extra.redis` (ends up [here](https://hypothesis.readthedocs.io/en/latest/database.html?highlight=redis.Redis#hypothesis.extra.redis.RedisExampleDatabase)).

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py (hunk 6 above)
Comment (sobolevn) [reply hypothesis-python/src/hypothesis/strategies/_internal/core.py:1026]: But, `typing_extensions` is not available during `coverage`. It is only used right now in `test_backported_types`:
Снимок экрана 2021-12-29 в 15 52 30

It is controlled by:

```bash
pip install 'typing_extensions>=4.0.0'
$PYTEST tests/typing_extensions/
if [ "$(python -c 'import sys; print(sys.version_info[:2] == (3, 7))')" = "False" ] ; then
  # Required by importlib_metadata backport, which we don't want to break
  pip uninstall -y typing_extensions
fi
```

Do you want me to change this?

Comment (sobolevn) [reply hypothesis-python/src/hypothesis/strategies/_internal/core.py:1026]: Another solution is to create python>=3.10 only file with `typing.TypeAlias` test.

File: hypothesis-python/scripts/basic-test.sh [hunk 10]
```diff
@@ -31,7 +31,7 @@ pip install fakeredis
 $PYTEST tests/redis/
 pip uninstall -y redis fakeredis
 
-pip install 'typing_extensions!=3.10.0.1'
+pip install 'typing_extensions>=4.0.0'
```
Comment (sobolevn) [inline hypothesis-python/scripts/basic-test.sh:34]: Do we need this version info?

**Comment by sobolevn** [comment]
@Zac-HD done! Only one doc failure is left.

**Review by Zac-HD — APPROVED** [review]
I pushed tiny patches for coverage and for the docs error 👍  Let's ship it!

**Comment by sobolevn** [comment]
Thank you for the review! I still have several extra types to forbid, I will send new PRs soon 🙂
